### PR TITLE
New general progress indication bar

### PR DESCRIPTION
**Pull request by talarian1**
- [x] All [tests](https://github.com/jfrog/jfrog-client-go#tests) passed. If this feature is not already covered by the tests, I added new tests.
- [x] This pull request is on the dev branch.
- [x] I used gofmt for formatting the code before submitting the pull request.
-----

### CODE REVIEW

File: httpclient/client.go [hunk 1]
```diff
@@ -195,11 +195,13 @@ func (jc *HttpClient) doUploadFile(localPath, url string, httpClientsDetails htt
...
 	reqContent := fileutils.GetUploadRequestContent(file)
 	var reader io.Reader
+	var ok = false
 	if file != nil && progress != nil {
-		progressId := progress.New(size, "Uploading", localPath)
-		reader = progress.ReadWithProgress(progressId, reqContent)
-		defer progress.Abort(progressId)
-	} else {
+		progressReader := progress.NewReaderProgressBar(size, "Uploading", localPath)
+		reader, ok = progressReader.ActionWithProgress(reqContent).(io.Reader)
```
Comment (RobiNino) [inline httpclient/client.go:201]: the ok check can probably be removed if ActionWithProgress is changed (here and in similar places)

**Review by RobiNino — CHANGES_REQUESTED** [review]
Looks great!

File: utils/io/progress.go [hunk 2]
```diff
@@ -1,19 +1,19 @@
...
+	// Initializes a new reader progress indication for a new file transfer.
+	// Input: 'total' - file size, 'prefix' - optional description, 'extraInformation' -extra information for disply.
 	// Output: progress indication id
-	New(total int64, prefix, filePath string) (id int)
+	NewReaderProgressBar(total int64, prefix, extraInformation string) (bar ProgressBar)
 	// Replaces an indication (with the 'replaceId') when completed. Used when an additional work is done as part of the transfer.
-	NewReplacement(replaceId int, prefix, filePath string) (id int)
-	// Used to wrap an io.Reader in order to track the bytes reading count of the file transfer, and update indication 'id' accordingly.
-	ReadWithProgress(id int, reader io.Reader) io.Reader
+	AddNewReplacementSpinner(replaceId int, prefix, extraInformation string) (id int)
```
Comment (eyalbe4) [inline utils/io/progress.go:10]: Let's change this API to a more generic. I suggest naming it *StateChanged*.

File: utils/io/progress.go [hunk 3]
```diff
@@ -1,19 +1,19 @@
...
+	// Returns the requested progress bar
+	GetProgressBar(id int) (bar ProgressBar)
```
Comment (eyalbe4) [inline utils/io/progress.go:12]: Let's rename to *GetProgress*.

File: utils/io/progress.go [hunk 4]
```diff
@@ -1,19 +1,19 @@
...
 	// Aborts a progress indication. Called on both successful and unsuccessful operations
 	Abort(id int)
```
Comment (eyalbe4) [inline utils/io/progress.go:14]: I suggest renaming *Abort(id int)* to *RemoveProgress(id int)*.

File: utils/io/progress.go [hunk 5]
```diff
@@ -1,19 +1,19 @@
...
 	// Quits the whole progress mechanism
 	Quit()
+	// Increases the general progress bar total count by given n.
```
Comment (eyalbe4) [inline utils/io/progress.go:17]: // Increases the general progress bar total count by given n. --> // Increments the general progress total count by given n.

File: utils/io/progressbar.go [hunk 6]
```diff
@@ -0,0 +1,12 @@
+package io
+
+import "io"
+
+type ProgressBar interface {
```
Comment (eyalbe4) [inline utils/io/progressbar.go:5]: I suggest moving this interface into *progress.go*, together with *ProgressMgr* intefface.
Let's also rename it to *Progress* and remove the word "Bar" from its comments and methods.

File: utils/io/progressbar.go [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+	// Uses to updated the progress bar progress.
+	ActionWithProgress(reader io.Reader) (results io.Reader)
+	// Aborts a progress indication. Called on both successful and unsuccessful operations
```
Comment (eyalbe4) [inline utils/io/progressbar.go:8]: indication --> indicator

**Review by eyalbe4 — CHANGES_REQUESTED** [review]

File: utils/io/progress.go [hunk 8]
```diff
@@ -3,17 +3,28 @@ package io
 import "io"
 
 // You may implement this interface to display progress indication of files transfer (upload / download)
-type Progress interface {
-	// Initializes a new progress indication for a new file transfer.
-	// Input: 'total' - file size, 'prefix' - optional description, 'filePath' - path of the file being transferred (for description purposes only).
+type ProgressMgr interface {
+	// Initializes a new reader progress indication for a new file transfer.
```
Comment (eyalbe4) [inline utils/io/progress.go:7]: indication --> indicator

File: utils/io/progress.go [hunk 9]
```diff
@@ -3,17 +3,28 @@ package io
...
+	// Aborts a progress indicator. Called on both successful and unsuccessful operations
+	RemoveProgress(id int)
 	// Quits the whole progress mechanism
 	Quit()
+	// Increments the general progress total count by given n.
+	IncGeneralProgressTotalBy(n int64)
+}
+
+type Progress interface {
+	// Uses to updated the progress indicator progress.
```
Comment (eyalbe4) [inline utils/io/progress.go:24]: Uses to updated... --> Used for updating...

**Review by eyalbe4 — CHANGES_REQUESTED** [review]

**Pull request merged**